### PR TITLE
feat(kubernetes events in orca): Exposes kubernetes events in orca for enhanced logging

### PR DESCRIPTION
**Pull request by chawlabhishek**
### :pencil: Description
- Exposes kubernetes events in orca that could be used for:
   - Providing more context in case of failures
   - Easier debugging of the issues

### :computer: Screenshots
Orca Events

### CODE REVIEW

**Comment by chawlabhishek** [comment]
> Can you provide an example showing where these events show up? I'm a little nervous that this could be a large amount of information that's going to be expensive to carry around.

Hi @dbyron-sf, 
Thank you for looking into the changes. The events get filtered as per cluster [coordinates](https://github.com/spinnaker/clouddriver/blob/3f4d2ab059feb9931e13d7cb05e4d6029b447011/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/view/provider/KubernetesManifestProvider.java#L71) in clouddriver. We'll get the resource specific events back in [orca](https://github.com/spinnaker/orca/blob/94cc3c301e95b8ae041aaf3ee29cb86896b7a255/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) depending on the includeEvents query parameter. Happy to attach some screenshots if it helps.
Thanks!

**Comment by dbyron-sf** [comment]
Yes please for some screenshots @chawlabhishek .

**Comment by chawlabhishek** [comment]
Hi @dbyron-sf,
Please find the attached events screenshot for a container bad count disruption. Please let me know if anything else is required. I'll be adding some dummy events for unit tests.
Thanks!

**Comment by dbyron-sf** [comment]
Thanks for that @chawlabhishek .  I was more thinking about snapshots of the spinnaker UI (include perhaps pipeline execution json) with the new information included.

**Comment by chawlabhishek** [comment]
> Thanks for that @chawlabhishek . I was more thinking about snapshots of the spinnaker UI (include perhaps pipeline execution json) with the new information included.

HI @dbyron-sf. Executed another experiment only to find out that this information does not get populated anywhere in spinnaker UI or pipeline execution json as of now. It has been made available to orca and could be easily switched off by setting the [includeEvents](https://github.com/spinnaker/orca/blob/94cc3c301e95b8ae041aaf3ee29cb86896b7a255/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) query parameter in getManifest as false.

File: orca-clouddriver/src/test/java/com/netflix/spinnaker/orca/clouddriver/model/ManifestTest.java [hunk 1]
```diff
@@ -69,6 +69,13 @@ void deserializesWarnings() throws IOException {
     assertThat(manifest.getWarnings()).isEmpty();
   }
 
+  @Test
+  void deserializesEvents() throws IOException {
+    String resource = getResource("clouddriver/model/manifests/stable.json");
```
[the remark this answers is not in the document]
Comment (chawlabhishek) [reply orca-clouddriver/src/test/java/com/netflix/spinnaker/orca/clouddriver/model/ManifestTest.java:74]: Validated with a  non-empty list of events.

**Comment by dbyron-sf** [comment]
> HI @dbyron-sf. Executed another experiment only to find out that this information does not get populated anywhere in spinnaker UI or pipeline execution json as of now. It has been made available to orca and could be easily switched off by setting the [includeEvents](https://github.com/spinnaker/orca/blob/94cc3c301e95b8ae041aaf3ee29cb86896b7a255/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) query parameter in getManifest as false.

I'm not sure I understand the point of the PR then....the title says "for enhanced logging", but...I don't see any log statements, and if the info doesn't go in the pipeline execution context or spinnaker UI...where does it go?

**Comment by chawlabhishek** [comment]
> I'm not sure I understand the point of the PR then....the title says "for enhanced logging", but...I don't see any log statements, and if the info doesn't go in the pipeline execution context or spinnaker UI...where does it go?

Hi @dbyron-sf, This information comes back in the manifest from clouddriver which was never pipeline execution context or spinnaker UI. Should someone wishes to use these events in orca, they can set the [includeEvents](https://github.com/spinnaker/orca/blob/94cc3c301e95b8ae041aaf3ee29cb86896b7a255/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) query parameter as true and then use them as they like. We have made it available in orca. Hope that makes sense.

**Comment by dbyron-sf** [comment]
It makes sense, but then I think we need a way to set includeEvents to true...either with some kind of config flag, or something controlled via information in a pipeline or...

**Comment by chawlabhishek** [comment]
> It makes sense, but then I think we need a way to set includeEvents to true...either with some kind of config flag, or something controlled via information in a pipeline or...

@dbyron-sf It could be easily set statically while calling the [getManifest](https://github.com/spinnaker/orca/blob/7fabdf8af372c6415fbbe1f2ba3d7a695261ebee/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) method. Do we want to make this configurable by any chance?

**Comment by chawlabhishek** [comment]
> Yes, it needs to be configurable, so people who don't want their pipeline execution to grow don't get the additional info.

@dbyron-sf That makes sense. Thank you.

**Comment by chawlabhishek** [comment]
> Yes, it needs to be configurable, so people who don't want their pipeline execution to grow don't get the additional info.

Hi @dbyron-sf, The scope of this PR was to make these events available in orca. The [includeEvents](https://github.com/spinnaker/orca/blob/94cc3c301e95b8ae041aaf3ee29cb86896b7a255/orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java#L89) parameter was being ignored in orca and the aim was to fix that.  Had a chat with @zrashwani on making the parameter configurable and as per our discussion a change would be required in multiple components (It's being used in multiple stages) and we believe that this should be addressed in a follow up PR. Could you please advise on the same?

**Comment by chawlabhishek** [comment]
> Yes, it needs to be configurable, so people who don't want their pipeline execution to grow don't get the additional info.

Hey @dbyron-sf, As discussed, includeEvents has been added as a configurable setting. It would be propagated from pipeline.json and if the value is not supplied, it would default to false. Could you review the change and provide your feedback on this PR?
FYI @zrashwani

File: orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java [hunk 2]
```diff
@@ -69,6 +69,7 @@ public TaskResult execute(@Nonnull StageExecution stage) {
     List<Map<String, String>> failedManifests = context.getFailedManifests();
     List warnings = context.getWarnings();
     List events = context.getEvents();
+    Boolean includeEvents = context.isIncludeEvents();
```
Comment (dbyron-sf) [inline orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java:72]: lower-case-b `boolean` is sufficient (and simpler) here ecause the type in WaitForManifestStableContext is lower-case-b `boolean`, right?
```suggestion
    boolean includeEvents = context.isIncludeEvents();
```

Comment (chawlabhishek) [reply orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/WaitForManifestStableTask.java]: @dbyron-sf Makes perfect sense. Updated.

**Comment by dbyron-sf** [comment]
@mergifyio update

**Review by dbyron-sf — APPROVED** [review]

**Comment by dbyron-sf** [comment]
@chawlabhishek would you mind adding a release note for this to https://github.com/spinnaker/spinnaker.io/blob/master/content/en/docs/releases/next-release-preview/index.md ?  Thanks!

**Comment by chawlabhishek** [comment]
> @chawlabhishek would you mind adding a release note for this to https://github.com/spinnaker/spinnaker.io/blob/master/content/en/docs/releases/next-release-preview/index.md ? Thanks!

@dbyron-sf Sure Thing!

**Pull request merged**